base/accumulate.jl

```diff
@@ -442,3 +459,22 @@ function _accumulate1!(op, B, v1, A::AbstractVector, dim::Integer)
     end
     return B
 end
+
+# Internal function used to identify the correct result array for a given op.
+# If that fails by returning `Union{}` then `nothing` will be returned, and the
+# calling function should fallback to an iterative solution.
+_try_result(args...; kwargs...) = nothing
+function _try_result(op, v::AbstractArray{T}; init=nothing) where T
+    # Try to identify the return type for the first element using `init` or `reduce_first`
+    F = isnothing(init) ? promote_op(reduce_first, typeof(op), T) : promote_op(op, typeof(init), T)
+    # Try to identify the return type fo the rest of the elements given op and eltype T
+    R = promote_op(op, T, T)
```

Comment:
```suggestion
    R = promote_op(op, F, T)
```
technically this still is wrong, since it only models the first and second arguments. The Stateful type has a similar problem in `_approx_iter_type` and you can see the approximate solution it employs there. The only place I know that handles this pattern fully complete is `@default_eltype`. I think the pattern you need here looks like this (untested):

```julia
function _accumulate_upper_bound(v, i, op, init)
    val = v[i]
    r = isnothing(init) ? reduce_first(op, val) : op(init, val)
    while inferencebarrier(nothing)
        val = v[i]
        r = op(r, val)
    end
    return r
end

function _try_result(op, v::AbstractArray{T}, init) where T
    return promote_op(_accumulate_upper_bound, typeof(v), Int, typeof(op), typeof(init))
end
```

No failure handling should be necessary (or desirable), since the only reason this returns `Union{}` is that the code will fail with the real error message before it reaches that assignment.

Reply:
> No failure handling should be necessary (or desirable), since the only reason this returns `Union{}` is that the code will fail with the real error message before it reaches that assignment.

Right, thanks! Since the original issue was that `reduce_first` would work and try to assign into the array of type `Union{}` when it should have been `Tuple{Int, Int}`. If `Union{}` is returned then it should mean that all operations prior to insertion should fail.

Hmm, so the edge case your snippet is trying to address is when the return type for each call to the `op` results in us calling a different method? My understanding is that we want a "mock" function which updates a result value with the `op` over all elements of the array. We don't call it at runtime, but pass it to `promote_op` so we can ensure the widest possible type? A couple things I don't understand are:

1. What does `inferencebarrier(nothing)` do here? I thought that was used to tell the compiler to infer a value as `Any`?
2. Would this slow down inference?

In any case, yes, your example does seem to address that edge case. Here's a gist showing the difference for that edge case.

https://gist.github.com/rofinn/6aae6f559f2d9e6d1539f063ceabe2fc

Reply:
Yes, it simply tells the compile to ignore that and treat the loop condition as unknown. Alternatively, this could perhaps use eachindex there to be a bit more accurate to the iteration protocol. There shouldn't be any discernible latency cost to this, since it is just one extra method to infer behavior on

I am also also not sure that typejoin result is needed on the uniontypes of the result. It looks like can be used directly in this example.

Reply:
Yeah, both work, but I was hoping to avoid overly complicated `Union` types.

Reply:
Okay, two other things I changed were.

1. Switched to using `Iterators.peel` in the inference function because tests failed for arrays that didn't support linear indexing.
2. I made it a nested function to make it clear that it is only intended to be used with `promote_op`.